### PR TITLE
RFC: Teach intersection that 4 != 5

### LINKED ISSUE #39088: Unreachable after Varargs change
```julia
a() = c((1,), (1,1,1,1))
c(d::NTuple{T}, ::NTuple{T}) where T = d
c(d::NTuple{f}, b) where f = c((d..., f), b)
j(h::NTuple{T}, ::NTuple{T} = a()) where T = nothing
j((1,1,1,1))
```

```
Unreachable reached at 0x7f5e68033107

signal (4): Illegal instruction
in expression starting at /home/tim/Julia/tools/creduce/main.jl:5
j at /home/tim/Julia/tools/creduce/main.jl:4
unknown function (ip: 0x7f5e68033160)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2232 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2414
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1708 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:117
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:206
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:157 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:563
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:670
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:879
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:827
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:931
eval at ./boot.jl:369 [inlined]
...
```

Bisected to #38136.

### PR DESCRIPTION
**Pull request by Keno**
This one is very tricky, because we need to represent
"Vararg with at least n-elements", which we don't really
have a representation for. This just normalizes it to
Tuple{Vararg{T, N}} and stores an additional constraint
out of line that `N` must be greater or equal `n`. Then
if `N` is not resolved to an integer, we will normalize
this in a post processing step to e.g. `Tuple{T, T, Vararg{T, N}}`
(for `n == 2`). Eventually it might make sense to give `Vararg`
itself a lower bound. For now this seems to work ok, but
this code is very tricky, so we should PkgEval it before
merging.

Fix #39088

### CODE REVIEW

File: src/simplevector.c [hunk 1]
```diff
@@ -81,6 +81,15 @@ JL_DLLEXPORT jl_svec_t *jl_svec_copy(jl_svec_t *a)
     return c;
 }
 
+JL_DLLEXPORT jl_svec_t *jl_svec_copy_resize(jl_svec_t *a, size_t n)
+{
+    size_t i, nold=jl_svec_len(a);
+    jl_svec_t *c = jl_alloc_svec_uninit(n);
+    for(i=0; i < nold && i < n; i++)
+        jl_svecset(c, i, jl_svecref(a,i));
+    return c;
```
Comment (vtjnash) [inline src/simplevector.c:90]:
```suggestion
    size_t i, nold = jl_svec_len(a);
    jl_svec_t *c = jl_alloc_svec_uninit(n);
    for (i = 0; i < nold && i < n; i++)
        jl_svecset(c, i, jl_svecref(a, i));
    for (; i < n; i++)
        jl_svecset(c, i, NULL);
    return c;
```

File: src/subtype.c [hunk 2]
```diff
@@ -2465,8 +2478,12 @@ static jl_value_t *finish_unionall(jl_value_t *res JL_MAYBE_UNROOTED, jl_varbind
...
     }
 
-    if (!varval && (vb->lb != vb->var->lb || vb->ub != vb->var->ub))
-        newvar = jl_new_typevar(vb->var->name, vb->lb, vb->ub);
+    int is_bound = 0;
+    if (!varval && vb->intvalued && jl_is_long(vb->lb) && vb->ub == (jl_value_t*)jl_any_type) {
+        is_bound = 1;
+        newvar = vb->lb;
+    } else if (!varval && (vb->lb != vb->var->lb || vb->ub != vb->var->ub))
+        newvar = (jl_value_t*)jl_new_typevar(vb->var->name, vb->lb, vb->ub);
```
Comment (vtjnash) [inline src/subtype.c:2486]:
```suggestion
    }
    else if (!varval && (vb->lb != vb->var->lb || vb->ub != vb->var->ub)) {
        newvar = (jl_value_t*)jl_new_typevar(vb->var->name, vb->lb, vb->ub);
    }
```

File: src/subtype.c [hunk 3]
```diff
@@ -2740,6 +2760,97 @@ static jl_value_t *intersect_varargs(jl_vararg_t *vmx, jl_vararg_t *vmy, jl_sten
...
+        assert(vb->offset > 0);
+        // Go backwards over the parameter list and verify that
+        // they are all equal to the vararg type.
+        for (ssize_t i = pos - 1; i > pos - 1 - vb->offset; i--) {
+            if (!forall_exists_equal(jl_svecref(*params, i), ii->T, e)) {
+                return 0;
+            }
+        }
+        jl_svecset(*params, pos - vb->offset, ii);
+        ((jl_svec_t*)*params)->length -= vb->offset;
```
Comment (vtjnash) [inline src/subtype.c:2788]: This can move an object from big to pooled, which may make the GC abort. Instead make a copy.

File: src/jltypes.c [hunk 4]
```diff
@@ -1480,32 +1497,64 @@ static jl_value_t *inst_tuple_w_(jl_value_t *t, jl_typeenv_t *env, jl_typestack_
...
+        jl_vararg_t *va = (jl_vararg_t *)jl_svecref(tp, ntp-1);
+        if (va->N) {
+            vaN = va->N;
+            if (jl_is_typevar(vaN)) {
+                jl_typeenv_t *te = find_typeenv(env, (jl_tvar_t*)vaN);
+                if (te) {
+                    if (te->val_is_lb) {
+                        assert(jl_is_long(te->val));
+                        nexpand = jl_unbox_long(te->val);
+                   } else {
```
Comment (ararslan) [inline src/jltypes.c:1513]:
```suggestion
                    } else {
```

File: src/simplevector.c [hunk 5]
```diff
@@ -81,6 +81,17 @@ JL_DLLEXPORT jl_svec_t *jl_svec_copy(jl_svec_t *a)
...
 }
 
+JL_DLLEXPORT jl_svec_t *jl_svec_copy_resize(jl_svec_t *a, size_t n)
+{
+    size_t i, nold=jl_svec_len(a);
+    jl_svec_t *c = jl_alloc_svec_uninit(n);
+    for(i=0; i < nold && i < n; i++)
+        jl_svecset(c, i, jl_svecref(a,i));
+    for (; i < n; i++)
+            jl_svecset(c, i, NULL);
```
Comment (ararslan) [inline src/simplevector.c:91]:
```suggestion
        jl_svecset(c, i, NULL);
```

**Comment by vtjnash** [comment]
I tried rebasing, but noted it currently has this particularly egregious error (both before and after rebasing):
`typeintersect(NTuple, Tuple{Any,Vararg}) === NTuple` (gives `Tuple{Any}`)

File: src/subtype.c [hunk 6]
```diff
@@ -2716,6 +2741,114 @@ static jl_value_t *intersect_varargs(jl_vararg_t *vmx, jl_vararg_t *vmy, jl_sten
...
+    ssize_t pos = jl_svec_len(*params) - 1;
+    if (vb->offset < 0) {
+        // Insert additional copies of the type to make up the
+        // `offset` difference.
+        *params = jl_svec_copy_resize(*params, jl_svec_len(*params) - vb->offset);
+        for (size_t i = 0; i < -vb->offset; ++i) {
+            jl_svecset(*params, pos + i, ii->T);
+        }
+        jl_svecset(*params, pos + -vb->offset, ii);
+    }
```
Comment (N5N3) [inline src/subtype.c:2759]: There seems no reason to do this.
If `vb->offset < 0`, these parameters should have been intersected during `intersect_tuple`.
A simple example
```julia
julia> A = Tuple{Tuple{T,Vararg{T,N}},NTuple{N,T}} where {N,T}
Tuple{Tuple{T, Vararg{T, N}}, Tuple{Vararg{T, N}}} where {N, T}

julia> T = Tuple{Tuple,NTuple{4,Int}} where {M}
Tuple{Tuple, NTuple{4, Int64}}

julia> typeintersect(A, T)
Tuple{NTuple{6, Int64}, NTuple{4, Int64}}
```
The above example wont be fixed after removing this branch, but that should be a bug during re-intersection. My local log shows this PR gives wrong result within the 1st round intersection.

Comment (N5N3) [reply src/subtype.c:2759]: Another bad regression.
```julia
julia> A = Tuple{Tuple{T,Vararg{T,N}},NTuple{N,T}} where {N,T}
Tuple{Tuple{T, Vararg{T, N}}, Tuple{Vararg{T, N}}} where {N, T}

julia> T = Tuple{Tuple,NTuple{M,Int}} where {M}
Tuple{Tuple, Tuple{Vararg{Int64, M}}} where M

julia> typeintersect(A, T)
Tuple{Tuple{Int64, Int64, Vararg{Int64, N}}, Tuple{}} where N
```
The first looks same as above.

Comment (vtjnash) [reply src/subtype.c:2759]: thanks for looking, I started this before I noticed your PR, and was looking at getting that merged instead, then to revisit whether this fixed any additional cases

**Comment by vtjnash** [comment]
Looks like all examples here are fixed now. And https://github.com/JuliaLang/julia/pull/46604 exists for continued accuracy improvements that might be possible.

**Pull request closed without merging**